test/e2e/scheduling/predicates.go

```diff
@@ -192,6 +194,135 @@ var _ = SIGDescribe("SchedulerPredicates [Serial]", func() {
 		verifyResult(cs, podsNeededForSaturation, 1, ns)
 	})
 
+	// This test verifies we don't allow scheduling of pods in a way that sum of limits +
+	// associated overhead is greater than machine's capacity.
+	// It assumes that cluster add-on pods stay stable and cannot be run in parallel
+	// with any other test that touches Nodes or Pods.
+	// Because of this we need to have precise control on what's running in the cluster.
+	// Test scenario:
+	// 1. Find the amount CPU resources on each node.
+	// 2. Create one pod with affinity to each node that uses 70% of the node CPU.
+	// 3. Wait for the pods to be scheduled.
+	// 4. Create another pod with no affinity to any node that needs 20% of the largest node CPU for the workload and
+	//    an overhead set as 30% of the largest node CPU.
+	// 5. Make sure this additional pod is not scheduled.
+
+	ginkgo.It("validates pod overhead is considered along with resource limits of pods that are allowed to run", func() {
+		WaitForStableCluster(cs, masterNodes)
+		nodeMaxAllocatable := int64(0)
+		nodeToAllocatableMap := make(map[string]int64)
+		for _, node := range nodeList.Items {
+			nodeReady := false
+			for _, condition := range node.Status.Conditions {
+				if condition.Type == v1.NodeReady && condition.Status == v1.ConditionTrue {
+					nodeReady = true
+					break
+				}
+			}
+			if !nodeReady {
+				continue
+			}
+			// Apply node label to each node
+			framework.AddOrUpdateLabelOnNode(cs, node.Name, "node", node.Name)
+			framework.ExpectNodeHasLabel(cs, node.Name, "node", node.Name)
+			// Find allocatable amount of CPU.
+			allocatable, found := node.Status.Allocatable[v1.ResourceCPU]
+			framework.ExpectEqual(found, true)
+			nodeToAllocatableMap[node.Name] = allocatable.MilliValue()
+			if nodeMaxAllocatable < allocatable.MilliValue() {
+				nodeMaxAllocatable = allocatable.MilliValue()
+			}
+		}
+		// Clean up added labels after this test.
+		defer func() {
+			for nodeName := range nodeToAllocatableMap {
+				framework.RemoveLabelOffNode(cs, nodeName, "node")
+			}
+		}()
+
+		pods, err := cs.CoreV1().Pods(metav1.NamespaceAll).List(metav1.ListOptions{})
```

Comment:
this can be _very_ slow.
can we filter based on some assumption?

Reply:
I'm not sure if this is really feasible. Again, this is following the existing resource checking test.

@bsalamat can you provide input here?

Reply:
one argument is that the suite should not be run on workload clusters, but we have no control over that if the users do it.

aren't field sectors an option here?
https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors/

e.g see here:
https://github.com/kubernetes/kubernetes/blob/acf64f450118cf46c8bdfa78d171d6059981af9d/test/e2e/apps/network_partition.go#L154-L155

Reply:
Our e2e tests run in real clusters which are non-hermetic. A good practice when writing these e2e tests is to assume that the cluster is running other workloads. So, if possible please try to list pods in the namespace of this test.